### PR TITLE
datetime: timezone support

### LINKED ISSUE #2886: Document timezone support
**Product:** Tarantool
**Since:** 
**Root document:**  https://www.tarantool.io/en/doc/latest/reference/reference_lua/
**SME:** @ ligurio  

Related to #2574 

# Details
Timezones support
-----------------

Tarantool uses IANA tzdata aka Olson DB facilities for timezone
resolution at the moment of parsing of datetime literals or while
parsing `tz` attribute in constructor, or `:set{}` method call.

In addition to the `tzoffset` we provided before, we now define
`tzindex`, the unique index assigned by Tarantool to each known IANA
timezone.

```lua
date = require 'datetime'

tarantool> T = date.parse '2022-01-01T00:00 Europe/Moscow'

tarantool> T.tzindex
---
- 947
...

tarantool> T.tzoffset
---
- 180
...

tarantool> T.tz
---
- Europe/Moscow
...
```

Now `date.isdst` field (alone and as part of `:totable()` table) is
correctly calculated using `tzindex` and attributes of the selected
timezone in the Olson DB timezone.

```
tarantool> date.parse('2004-06-01T00:00 Europe/Moscow').isdst
---
- true
...

tarantool> date.parse('2004-12-01T00:00 Europe/Moscow'):totable()
---
- sec: 0
  min: 0
  yday: 336
  day: 1
  nsec: 0
  isdst: false
  wday: 4
  tzoffset: 180
  month: 12
  year: 2004
  hour: 0
...

```

All timezone names and abbreviations, known to the current Tarantool
version are available via `datetime.TZ` bidirectional array.

```lua
print(date.TZ['Europe/Moscow']) -- 947
print(date.TZ[947]) -- Europe/Moscow
```

Limitations
-----------

There were moments in past history, when local mean time in some
partcular zone used timezone offset not representable in whole
minutes, but rather in seconds, i.e. in Moscow before 1918 there
used to be offset +2 hours 31 minutes and 19 seconds. Please see
Olson dump for this period

```
$ ./src/lib/tzcode/install/usr/bin/zdump -c1880,1918 -i Europe/Moscow

TZ="Europe/Moscow"
-       -       +023017 MMT
1916-07-03      00:01:02        +023119 MMT
1917-07-02      00      +033119 MST     1
1917-12-27      23      +023119 MMT
```

Modern tzdata rules do not use such tiny fraction, and all timezones differ
to UTC in units measured in minutes, not seconds. Tarantool datetime uses
minutes internally as units for `tzoffset` so there is some loss of
precision if you try to operate with such ancient timestamps.
Requested by @ tsafin in https://github.com/tarantool/tarantool/commit/979a8fbd5a4f4cd087e6f1aed6aeb3cb690cdb5d.

### PR DESCRIPTION
**Pull request by ligurio**
Stage: https://docs.d.tarantool.io/en/doc/gh-2886-timezone-support/reference/reference_lua/datetime/

Closes #2886

### CODE REVIEW

File: doc/reference/reference_lua/datetime.rst [hunk 1]
```diff
@@ -155,12 +155,14 @@ Functions
                 -   0
 
             *   -   tzoffset
-                -   Time zone offset from UTC, in minutes. If both ``tzoffset`` and ``tz`` are specified, ``tz`` has the preference and the ``tzoffset`` value is ignored.
+                -   Time zone offset from UTC, in minutes. If both ``tzoffset`` and
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
                -   A time zone offset from UTC, in minutes. If both ``tzoffset`` and
```

File: doc/reference/reference_lua/datetime.rst [hunk 2]
```diff
@@ -155,12 +155,14 @@ Functions
...
+                    ``tz`` are specified, ``tz`` has the preference and the ``tzoffset``
+                    value is ignored. See a section :ref:`timezone <timezone>`.
                 -   number
                 -   0
 
             *   -   tz
-                -   Time zone name according to the `tz database <https://en.wikipedia.org/wiki/Tz_database>`__.
+                -   Time zone name according to the Time Zone Database, see a section :ref:`timezone <timezone>`.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
                -   A time zone name according to the Time Zone Database. See the :ref:`timezone` section.
```

File: doc/reference/reference_lua/datetime.rst [hunk 3]
```diff
@@ -497,9 +505,21 @@ Properties
 
     Since: :doc:`2.11.0 </release/2.11.0>`
 
-    An array of timezone names and abbreviations.
+    A Lua table that maps timezone names (like ``Europe/Moscow``) and
+    timezone abbreviations (like ``MSK``) to its index and vice-versa,
+    see a section :ref:`timezone <timezone>`.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
    timezone abbreviations (like ``MSK``) to its index and vice-versa.
    See the :ref:`timezone` section.
```

File: doc/reference/reference_lua/datetime.rst [hunk 4]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
         ...
 
+.. _timezone:
+
+Time zones
+----------
+
+Full support has been added since :doc:`2.11.0 </release/2.11.0>`.
+
+Tarantool uses `Time Zone Database <https://www.iana.org/time-zones>`__
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
Tarantool uses the `Time Zone Database <https://www.iana.org/time-zones>`__
```

File: doc/reference/reference_lua/datetime.rst [hunk 5]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+(also known as Olson database and supported by IANA) for timezone support.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
(also known as the Olson database and supported by IANA) for timezone support.
```

File: doc/reference/reference_lua/datetime.rst [hunk 6]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+You can use the Lua module :ref:`tarantool <tarantool-module>` to get a used version of ``tzdata``.
+
+Every datetime object has three fields that represents timezone support:
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
Every datetime object has three fields that represent timezone support:
```

Or we can make it shorter: `fields that represent timezone support -> fields for timezone support`.

File: doc/reference/reference_lua/datetime.rst [hunk 7]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+        tarantool> require('datetime').parse('2004-06-01T00:00 Europe/Moscow').isdst
+        ---
+        - true
+        ...
+
+*   The field ``tz`` field can be set to timezone name or timezone abbreviation. Timezone name
+    is a human-readable timezone names, basing on the Time Zone Database, for example "Europe/Moscow".
+    Timezone abbreviation represents time zones by `alphabetic abbreviations <https://www.timeanddate.com/time/zones/>`__
+    such as "EST", "WST", and "F". Both timezone names and abbreviations are available
+    via bidirectional array :ref:`datetime.TZ <datetime-tz>`.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
*   The field ``tz`` field can be set to a timezone name or abbreviation. A timezone name
    is a human-readable name based on the Time Zone Database, for example, "Europe/Moscow".
    Timezone abbreviations represents time zones by `alphabetic abbreviations <https://www.timeanddate.com/time/zones/>`__
    such as "EST", "WST", and "F". Both timezone names and abbreviations are available
    via the bidirectional array :ref:`datetime.TZ <datetime-tz>`.
```

File: doc/reference/reference_lua/datetime.rst [hunk 8]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+
+*   The field ``tzoffset`` is calculated automatically using current Olson rule i.e.
+    it would take into consideration summer time, leap year and leap seconds information
+    when timezone name is set. However, field ``tzoffset`` can be set manually when
+    appropriate timezone is not available.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
*   The field ``tzoffset`` is calculated automatically using the current Olson rule.
    This means that it takes into account summer time, leap year, and leap seconds information
    when a timezone name is set. However, the ``tzoffset`` field can be set manually when
    an appropriate timezone is not available.
```

File: doc/reference/reference_lua/datetime.rst [hunk 9]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+
+The fields ``tz``, ``tzoffset`` can be set :ref:`datetime.new() <datetime-new>`,
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
The fields ``tz``, ``tzoffset`` can be set in :ref:`datetime.new() <datetime-new>`,
```

File: doc/reference/reference_lua/datetime.rst [hunk 10]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+:ref:`datetime.parse() <datetime-parse>` and :ref:`datetime_object:set() <datetime-set>`.
+
 Limitations
 -----------
 
-The supported date range is from ``-5879610-06-22`` to ``+5879611-07-11``.
+*   The supported date range is from ``-5879610-06-22`` to ``+5879611-07-11``.
+
+*   There were moments in a past history, when local mean time in some partcular zone
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
*   There were moments in past history when local mean time in some particular zone
```

File: doc/reference/reference_lua/datetime.rst [hunk 11]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+    used timezone offset not representable in a whole minutes, but rather in seconds,
+    i.e. in Moscow before 1918 there used to be offset +2 hours 31 minutes and 19 seconds.
+    See an Olson dump for this period:
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
    used a timezone offset not representable in a whole minutes but rather in seconds.
    For. example, in Moscow before 1918 there used to be offset +2 hours 31 minutes and 19 seconds.
    See an Olson dump for this period:
```

File: doc/reference/reference_lua/datetime.rst [hunk 12]
```diff
@@ -1080,10 +1110,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+        TZ="Europe/Moscow"
+        -       -       +023017 MMT
+        1916-07-03      00:01:02        +023119 MMT
+        1917-07-02      00      +033119 MST     1
+        1917-12-27      23      +023119 MMT
+
+    Modern ``tzdata`` rules do not use such tiny fraction, and all timezones differ
+    to UTC in units measured in minutes, not seconds. Tarantool datetime module uses
+    minutes internally as units for ``tzoffset`` so there is some loss of precision
+    if you try to operate with such ancient timestamps.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
    Modern ``tzdata`` rules do not use such a tiny fraction, and all timezones differ
    from UTC in units measured in minutes, not seconds. Tarantool datetime module uses
    minutes internally as units for ``tzoffset``. So there might be some loss of precision
    if you try to operate with such ancient timestamps.
```

**Review by andreyaksenov — CHANGES_REQUESTED** [review]
Please see my comments below:

**Comment by ligurio** [comment]
Added:

```diff
diff --git a/doc/reference/reference_lua/datetime.rst b/doc/reference/reference_lua/datetime.rst
index 32401669..292a0aef 100644
--- a/doc/reference/reference_lua/datetime.rst
+++ b/doc/reference/reference_lua/datetime.rst
@@ -541,7 +541,8 @@ datetime_object
 
         Modify an existing datetime object by adding values of the input argument.
         See also: :ref:`interval_arithm`. The addition is performed taking ``tzdata``
-        into account ``tzoffset`` or ``tz`` fields are set.
+        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`
+        and the :ref:`interval_arithm` sections.
 
         :param table input: an :ref:`interval object <interval_obj>` or an equivalent table (see **Example #1**)
         :param string adjust: defines how to round days in a month after an arithmetic operation.
@@ -723,7 +724,8 @@ datetime_object
 
         Modify an existing datetime object by subtracting values of the input argument.
         See also: :ref:`interval_arithm`. The subtraction is performed taking ``tzdata``
-        into account ``tzoffset`` or ``tz`` fields are set.
+        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`
+        and the :ref:`interval_arithm` sections.
 
         :param table input: an :ref:`interval object <interval_obj>` or an equivalent table (see **Example**)
         :param string adjust: defines how to round days in a month after an arithmetic operation.
@@ -1021,6 +1023,16 @@ The matrix of the ``subtraction`` operands eligibility and their result types:
             -   interval
             -   interval
 
+The subtraction and addition of datetime objects are performed taking ``tzdata``
+into account ``tzoffset`` or ``tz`` fields are set:
+
+..  code-block:: tarantoolsession
+
+    tarantool> datetime.new({tz='MSK'}) - datetime.new({tz='UTC'})
+    ---
+    - -180 minutes
+    ...
+
 .. _interval_comp:
 
 Datetime and interval comparison
```

File: doc/reference/reference_lua/datetime.rst [hunk 13]
```diff
@@ -1080,10 +1120,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+----------
+
+Full support has been added since :doc:`2.11.0 </release/2.11.0>`.
+
+Tarantool uses the `Time Zone Database <https://www.iana.org/time-zones>`__
+(also known as the Olson database and supported by IANA) for timezone support.
+You can use the Lua module :ref:`tarantool <tarantool-module>` to get a used version of ``tzdata``.
+
+Every datetime object has three fields that represent timezone support:
+``tz``, ``tzoffset`` and ``isdst``:
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
``tz``, ``tzoffset``, and ``isdst``:
```

File: doc/reference/reference_lua/datetime.rst [hunk 14]
```diff
@@ -1080,10 +1120,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+    ..  code-block:: tarantoolsession
+
+        tarantool> require('datetime').parse('2004-06-01T00:00 Europe/Moscow').isdst
+        ---
+        - true
+        ...
+
+*   The field ``tz`` field can be set to a timezone name or abbreviation. A timezone name
+    is a human-readable name based on the Time Zone Database, for example, "Europe/Moscow".
+    Timezone abbreviations represents time zones by `alphabetic abbreviations <https://www.timeanddate.com/time/zones/>`__
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
    Timezone abbreviations represent time zones by `alphabetic abbreviations <https://www.timeanddate.com/time/zones/>`__
```

File: doc/reference/reference_lua/datetime.rst [hunk 15]
```diff
@@ -1080,10 +1120,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+    such as "EST", "WST", and "F". Both timezone names and abbreviations are available
+    via the bidirectional array :ref:`datetime.TZ <datetime-tz>`.
+
+*   The field ``tzoffset`` is calculated automatically using the current Olson rule.
+    This means that it takes into account summer time, leap year, and leap seconds information
+    when a timezone name is set. However, the ``tzoffset`` field can be set manually when
+    an appropriate timezone is not available.
+
+The fields ``tz``, ``tzoffset`` can be set in :ref:`datetime.new() <datetime-new>`,
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
The fields ``tz`` and ``tzoffset`` can be set in :ref:`datetime.new() <datetime-new>`,
```

File: doc/reference/reference_lua/datetime.rst [hunk 16]
```diff
@@ -1080,10 +1120,68 @@ This section describes how the ``datetime`` module supports leap seconds:
...
+:ref:`datetime.parse() <datetime-parse>` and :ref:`datetime_object:set() <datetime-set>`.
```
Comment (andreyaksenov) [inline doc/reference/reference_lua/datetime.rst]:
```suggestion
:ref:`datetime.parse() <datetime-parse>`, and :ref:`datetime_object:set() <datetime-set>`.
```

**Comment by ligurio** [comment]
Updated:

```diff
diff --git a/doc/reference/reference_lua/datetime.rst b/doc/reference/reference_lua/datetime.rst
index 292a0aef..d785fe83 100644
--- a/doc/reference/reference_lua/datetime.rst
+++ b/doc/reference/reference_lua/datetime.rst
@@ -541,8 +541,7 @@ datetime_object
 
         Modify an existing datetime object by adding values of the input argument.
         See also: :ref:`interval_arithm`. The addition is performed taking ``tzdata``
-        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`
-        and the :ref:`interval_arithm` sections.
+        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`.
 
         :param table input: an :ref:`interval object <interval_obj>` or an equivalent table (see **Example #1**)
         :param string adjust: defines how to round days in a month after an arithmetic operation.
@@ -724,8 +723,7 @@ datetime_object
 
         Modify an existing datetime object by subtracting values of the input argument.
         See also: :ref:`interval_arithm`. The subtraction is performed taking ``tzdata``
-        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`
-        and the :ref:`interval_arithm` sections.
+        into account, when ``tzoffset`` or ``tz`` fields are set, see the :ref:`timezone`.
 
         :param table input: an :ref:`interval object <interval_obj>` or an equivalent table (see **Example**)
         :param string adjust: defines how to round days in a month after an arithmetic operation.
@@ -1157,6 +1155,8 @@ Every datetime object has three fields that represent timezone support:
 
 The fields ``tz``, ``tzoffset`` can be set in :ref:`datetime.new() <datetime-new>`,
 :ref:`datetime.parse() <datetime-parse>` and :ref:`datetime_object:set() <datetime-set>`.
+The arithmetic on datetime objects are performed taking ``tzdata`` into account, when
+``tzoffset`` or ``tz`` fields are set, see the :ref:`interval_arithm` section.
 
 Limitations
 -----------
```

**Review by andreyaksenov — CHANGES_REQUESTED** [review]
Thanks! A couple of minor things

**Review by andreyaksenov — APPROVED** [review]
Thanks! Merging

**Pull request merged**